File: TREEORD.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def buildTree(inOrd, preOrd, inStr, inEnd):
    if (inStr > inEnd):
        return None

    tNode = Node(preOrd[buildTree.preIndex])
    buildTree.preIndex += 1

    if inStr is inEnd:
        return tNode

    inIndex = search(inOrd, inStr,  inEnd, tNode.data)

    tNode.left = buildTree(inOrd, preOrd, inStr, inIndex - 1)
    tNode.right = buildTree(inOrd, preOrd, inIndex + 1, inEnd)

    return tNode


def search(arr, start, end, value):
    for i in range(start, end+1):
        if arr[i] == value:
            return i


def printInorder(node):
    if node is None:
        return

    printInorder(node.left)
    print(node.data, end = " ")
    printInorder(node.right)


def postorder(node):
    if node is None:
        return

    postorder(node.left)
    postorder(node.right)
    # print(node.data, end = " ")
    posVec.append(node.data)


def cospe():
    if ((posVec > posOrder) - (posVec < posOrder)) is 0:
        print('yes\n')
    else:
        print('no\n')

if __name__ == '__main__':
#     preOrder = ['1', '2', '4', '5', '3', '6']
#     posOrder = ['4', '5', '2', '6', '3', '1']
#     inOrder = ['4', '2', '5', '1', '3', '6']

    # preOrder = ['1', '2', '4', '5', '3', '6']
    # posOrder = ['4', '5', '2', '6', '1', '3']
    # inOrder = ['4', '2', '5', '1', '6', '3']

    posVec = []

    nos = int(input())
    preOrder = input().split(' ')
    posOrder = input().split(' ')
    inOrder = input().split(' ')

    # print(preOrder, posOrder, inOrder, sep = '\n')

    buildTree.preIndex = 0
    while 1:
        try:
            root = buildTree(inOrder, preOrder, 0, nos-1)
            postorder(root)
            cospe()
            break
        except:
            print('no\n')
            break

    # print(posVec is posOrder)
    # print(((posVec > posOrder) - (posVec < posOrder)))

    # print("Inorder traversal of the constructed tree is")
    # printInorder(root)
    # print('')
    #
    # postorder(root)
    # print('')

File: TREEORD.java
import java.io.*;

class Main {
	static class No{
		int data;
		No left;
		No right;

		public No(int data){
      this.data = data;
      left = null;
      right = null;
		}
	}

	static class Arvore{
		protected String posStr = "";
		protected static int preIndex = 0;

		private No monta(String in[], String pre[], int inStrt, int inEnd){
      if (inStrt > inEnd)
        return null;

      int data = Integer.parseInt(pre[preIndex++]);
      No tNo = new No(data);

      if (inStrt == inEnd)
        return tNo;

      int inIndex = search(in, inStrt, inEnd, tNo.data);

      tNo.left = monta(in, pre, inStrt, inIndex - 1);
      tNo.right = monta(in, pre, inIndex + 1, inEnd);

      return tNo;
	    }

		private int search(String vec[], int comeca, int fim, int data){
      int i;
      for (i = comeca; i <= fim; i++){
      	int vecValue = Integer.parseInt(vec[i]);
          if ( vecValue == data)
            return i;
      }
      return i;
    }

		private void posOrdemArvore(No no){
      if (no == null) return;

      posOrdemArvore(no.left);
      posOrdemArvore(no.right);
      posStr += no.data + " ";
    }

		private String equalsStr() {
			return posStr;
		}

	}

	public static void main(String[] args) throws IOException{
    Arvore tree = new Arvore();

    BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
    int tamanho = Integer.parseInt(in.readLine());

    String pre = in.readLine();
    String preOrdem[] = pre.split(" ");

    String pos = in.readLine();

    String inOr = in.readLine();
    String inOrdem[] = inOr.split(" ");

    No root = tree.monta(inOrdem, preOrdem, 0, tamanho-1);

    tree.posOrdemArvore(root);

    String posArvore = tree.equalsStr();

    String posStr = posArvore.substring(0, posArvore.length() - 1);
    if(pos.equals(posStr)) {
            System.out.println("yes");
    }

    else {
            System.out.println("no");
    }
	}
}
